### ISSUE TITLE
Avoiding repo sprawl because of Python components

### ISSUE THREAD

**Issue by ghost** [openSUSE/daps#632]
So I'd already mentioned that I think this strategy of creating new repos for each of our small Python scripts and manually copying these scripts when need arises is a bit weird, and leads to blobby, barely reviewable commits which are rather less suitable for longer-term maintenance of DAPS. It also makes it harder to learn about where you should actually commit stuff into.

I think this issue could be resolved with a combination of symlinks + integrating the GHA workflows into the main `.github/workflows` dir:

```
daps
├── .github
│   ├── [issue template etc.]
│   └── workflows
│       ├── ci.yml
│       ├── python-xmlwellformed.yml
│       └── python-getentities.yml
├── python_scripts
│   ├── daps-xmlwellformed
│   │   └── [contents of the daps-xmlwellformed repo]
│   └── getentities
│       └── [contents of the getentities repo]
├── libexec
│   ├── daps-xmlwellformed --symlink--> ../python_scripts/daps-xmlwellformed/bin/daps-xmlwellformed.py
│   ├── getentities        --symlink--> ../python_scripts/getentities/bin/getentities.py
│   └── [other contents of the libexec/]
│
[... all the rest of the DAPS repo ...]
```

The remaining question would be how to package this -- imo the `python_scripts` dir should not be part of the RPM, rather we should resolve the symlinks from the `libexec` dir.

@tomschr / @fsundermeyer What do you think?

**Comment by tomschr** [comment]
Thanks Stefan. We briefly touched this topic in our sync call.

I see your point, however I don't think your concerns are that problematic because:

a. these scripts don't change much and are unlikely to in the future.
b. maintenance effort is quite low.
c. copying a script from one repo to another is not ideal, but if you add a good log message with commits from the other repo etc. I see it as a good compromise. However, because of a) and b) this will only happen if we find (new) bugs or need to add new features (unlikely).
d. separate repos avoid unnecessary files from the daps repo.

**For me** it was the easiest way to separate tests without bloating the daps repo with unwanted files. It **does not** mean that I'll move every little script in its own repo. The two mentioned scripts are some crucial parts in the daps workflow that needed special care unfortunately.

If you are really concerned about commits, maybe we need to collect all the pros and cons and look into submodules/subtrees. I know that this will not convince you. :wink: I also believe, this will be going overboard.

I've searched the Internet a bit and there seems to be no general consensus *when* to move one part into a separate repo. Some use a mono/single repo, others a more modular approach. I found the [Choosing between Single or multiple projects in a git repository?](https://softwareengineering.stackexchange.com/q/161293) quite interesting (lists the pros and cons).

**TL;DR:** Tom uses the pragmatic approach of using separate repos for the two scripts, because they won't change much, have low maintenance efforts, avoids a bloated daps repo, and separate concerns.

**Comment by ghost** [comment]
> a. these scripts don't change much and are unlikely to in the future.
> b. maintenance effort is quite low.

That's the same thing twice and I don't see these as arguments for or against anything. I also don't think they're true, given how you mention that they are "crucial parts in the daps workflow" and given that the reason you've been outsourcing them is to be able to fix issues in both scripts.


> c. copying a script from one repo to another is not ideal, but if you add a good log message with commits from the other repo etc. I see it as a good compromise.

I see it as a terrible compromise. Especially when we already have commits with completely misleading commit messages like this: https://github.com/openSUSE/daps/commit/2656c7fb241f3133d44defad1a4193d102d1f568 (NB: This commit did nothing to address the bug mentioned in the message.)


> If you are really concerned about commits, maybe we need to collect all the pros and cons and look into submodules/subtrees. I know that this will not convince you. wink I also believe, this will be going overboard.

Indeed, no necessity to drag those complications in.

> d. separate repos avoid unnecessary files from the daps repo.

This makes my head hurt.

* DAPS depends on daps-xmlwellformed and getentities. DAPS is the only software in the world that depends on these two scripts. And yet, you argue that making sure you can develop these scripts as part of DAPS "bloats the repo"?
* We also have tons of stuff within daps-xslt/ that is absolutely not needed by DAPS, is not on the way toward integration, and yet it is argued that we need to ship it, https://github.com/openSUSE/daps/issues/612

**Comment by tomschr** [comment]
>  you mention that they are "crucial parts in the daps workflow" and given that the reason you've been outsourcing them is to be able to fix issues in both scripts.

IMHO, I don't think it's a contradiction. As I said, it won't change that much. If Frank is happy with my current implementation, we can commit it with a good log message. Case closed, problem solved. I don't think they will see lots of commits in the near future.

> Especially when we already have commits with completely misleading commit messages like this: 2656c7f

Bad things can always happen, regardless of which approach you follow.


> DAPS depends on daps-xmlwellformed and getentities. DAPS is the only software in the world that depends on these two scripts. And yet, you argue that making sure you can develop these scripts as part of DAPS "bloats the repo"?

Sure, the scripts are tiny. It was one aspect. You can argue if "bloating" is the correct linguistic term. As a matter of fact, it makes the daps repo bigger with files that are not really needed. I prefer them to be separated, similar what you did with configs for docserv.

> We also have tons of stuff within daps-xslt/ that is absolutely not needed by DAPS, [...] and yet it is argued that we need to ship it

Not sure what this brings to the actual discussion. :man_shrugging: 

Let's summarize this: You have a different approach than me. :slightly_smiling_face: That's fine. Let's wait and discuss this in our next doc sync call.

**Comment by ghost** [comment]
> similar what you did with configs for docserv.

Not sure how to interpret what you're saying here, I'm assuming you're talking about the Docserv2 repo versus the docserv-config repo. But that would be a terrible comparison. Docserv2 is the generic, public tool. docserv-config holds our specialized, private configuration for our instance of the tool.

>Not sure what this brings to the actual discussion. man_shrugging

It shows the logical inconsistency in the argument.

**Comment by tomschr** [comment]
As decided in our last sync meeting (July 29th):

* Move content over from my tomschr repo into daps repo.
* Add GitHub Action workflow which is only triggered when specific paths or Python files have been changed.
* Make it possible to allow testing of different Python scripts.

I guess I need some ideas about the structure and which files should go into which directories.

Assigned to me.

**Comment by tomschr** [comment]
It's implemented now, so closing this issue.
